src/pyuvdata/uvbeam/analytic_beam.py

```diff
@@ -0,0 +1,893 @@
+from typing import Literal
+
+import numpy as np
+import numpy.typing as npt
+import yaml
+from astropy.constants import c as speed_of_light
+from scipy.special import j1
+
+from .. import utils
+from ..docstrings import combine_docstrings
+from .uvbeam import UVBeam, _convert_feeds_to_pols
+
+analytic_beam_classes = ["AiryBeam", "GaussianBeam", "ShortDipoleBeam", "UniformBeam"]
+__all__ = ["AnalyticBeam"] + analytic_beam_classes
+
+
+@dataclass
+class AnalyticBeam(ABC):
+    """
+    Define an analytic beam base class.
+
+    Attributes
+    ----------
+    basis_vec_dict : dict
+        Allowed basis vector types (keys) with their number of axes (which will be set
+        on the object as Naxes_vec).
+    basis_vector_type : str
+        The type of basis vectors this beam is defined with. Must be a key in
+        `basis_vec_dict`. Currently, only "az_za" is supported, which means basis
+        vectors aligned with the zenith angle and azimuth.
+    Naxes_vec : int
+        Number of basis vectors. This is set based on the `basis_vector_type`.
+    Nfeeds : int
+        Number of feeds. This is set based on the `feed_array`.
+
+    Parameters
+    ----------
+    feed_array : np.ndarray of str
+        Feeds to define this beam for, e.g. x & y or n & e (for "north" and "east")
+        or r & l.
+    include_cross_pols : bool
+        Option to include the cross polarized beams (e.g. xy and yx or en and ne) for
+        the power beam.
+    x_orientation : str
+        Physical orientation of the feed for the x feed. Not meaningful for
+        unpolarized analytic beams, but clarifies the orientation of the dipole
+        for the ShortDipoleBeam and matches with the meaning on UVBeam objects.
+
+    """
+
+    # In the future, this might allow for cartesian basis vectors in some orientation.
+    # In that case, the Naxes_vec would be 3 rather than 2
+    _basis_vec_dict = {"az_za": 2}
+
+    @property
+    @abstractmethod
+    def basis_vector_type(self):
+        """Require that a basis_vector_type is defined in concrete classes."""
+
+    def __init__(
+        self,
+        *,
+        feed_array: npt.NDArray[str] | None = None,
+        include_cross_pols: bool = True,
+        x_orientation: Literal["east", "north"] = "east",
+    ):
+        if self.basis_vector_type not in self._basis_vec_dict:
+            raise ValueError(
+                f"basis_vector_type is {self.basis_vector_type}, must be one of "
+                f"{list(self._basis_vec_dict.keys())}"
+            )
+        self.Naxes_vec = self._basis_vec_dict[self.basis_vector_type]
+
+        if feed_array is not None:
+            for feed in feed_array:
+                allowed_feeds = ["n", "e", "x", "y", "r", "l"]
+                if feed not in allowed_feeds:
+                    raise ValueError(f"Feeds must be one of: {allowed_feeds}")
+        else:
+            feed_array = ["x", "y"]
+        self.feed_array = np.asarray(feed_array)
```

Comment:
Again, this can be managed in the `__post_init__`, but if we were using `attrs` (or `pydantic`, though I haven't used that) it would be even simpler:

```python
import attrs
from attrs import validators as vld

@attrs.define
class AnalyticBeam(ABC):
    ...
    feed_array: np.ndarray = attrs.field(
        converter=np.asarray, 
        validator=vld.deep_iterable(vld.in_(["n", "e", "x", "y", "r", "l"]))
    )